third_party/amd/lib/TritonAMDGPUToLLVM/LoadStoreOpToLLVM.cpp

```diff
@@ -98,23 +103,109 @@ struct LoadStoreConversionBase {
+    Value v = undef(vecTy);
+    for (size_t s = 0; s < vec; ++s) {
+      Value otherElem = elems[start + s];
+      Value indexVal =
+          LLVM::createIndexConstant(rewriter, loc, typeConverter, s);
+      v = insert_element(vecTy, v, otherElem, indexVal);
+    }
+    return v;
+  }
+
+  // Return a tensor of pointers with the same type of `basePtr` and the same
+  // shape of `offset`
+  Type getPointerTypeWithShape(Value basePtr, Value offset) const {
+    Type basePtrType = basePtr.getType();
+    auto offsetType = cast<RankedTensorType>(offset.getType());
+    return RankedTensorType::get(offsetType.getShape(), basePtrType,
+                                 offsetType.getEncoding());
+  }
+
+  // Get contiguity for a tensor pointer `ptr`
   unsigned getContiguity(Value ptr) const {
     auto tensorTy = dyn_cast<RankedTensorType>(ptr.getType());
     if (!tensorTy)
       return 1;
     return axisAnalysisPass.getPtrContiguity(ptr);
   }
 
+  // Get contiguity for a scalar pointer `ptr` and a tensor `offset`
+  unsigned getContiguity(Value ptr, Value offset) const {
+    // Get contiguity from the offset
+    Type type = getPointerTypeWithShape(ptr, offset);
+    RankedTensorType tensorTy = cast<RankedTensorType>(type);
+    auto layout = tensorTy.getEncoding();
+    auto order = triton::gpu::getOrder(layout);
+    auto uniqueContigPerThread =
+        triton::gpu::getUniqueContigPerThread(layout, tensorTy.getShape());
+    assert(order[0] < uniqueContigPerThread.size() &&
+           "Unexpected uniqueContigPerThread size");
+    unsigned contiguity = uniqueContigPerThread[order[0]];
+
+    // Get alignment from the pointer. Since this is a scalar pointer
+    // we should not take the pointer contiguity to consider alignment
+    auto *axisInfo = axisAnalysisPass.getAxisInfo(ptr);
+    auto maxMultipleBytes = axisInfo->getDivisibility(order[0]);
+    auto elemNumBits = triton::getPointeeBitWidth(tensorTy);
+    auto elemNumBytes = std::max<unsigned>(elemNumBits / 8, 1);
+    auto align = std::max<int64_t>(maxMultipleBytes / elemNumBytes, 1);
+
+    // Final contiguity is a min of the offset contiguity and pointer alignment
+    contiguity = std::min<int64_t>(align, contiguity);
+    return contiguity;
+  }
+
+  // Determine the vector size of a tensor of pointers
   unsigned getVectorSize(Value ptr) const {
     auto tensorTy = dyn_cast<RankedTensorType>(ptr.getType());
     if (!tensorTy)
       return 1;
     auto contiguity = getContiguity(ptr);
     auto pointeeBitWidth = triton::getPointeeBitWidth(tensorTy);
-    // The maximum vector size is 128 bits on NVIDIA GPUs.
     return std::min<unsigned>(128 / pointeeBitWidth, contiguity);
   }
 
+  // Given a scalar pointer and a tensor of offsets, determine the vector size
+  unsigned getVectorSize(Value ptr, Value offset) const {
+    auto contiguity = getContiguity(ptr, offset);
+    auto pointeeBitWidth = triton::getPointeeBitWidth(ptr.getType());
```

Comment:
Do we need to adjust for `i1` here?

Reply:
If I understand what you mean, I don't think we need. I mean, we are doing exactly what the normal `getVectorSize(ptr)` is doing, except we need to account for the fact that now the type needs to be extracted from `ptr` and the shape from the `offset`. If we want to account for a pointer to `i1` (assuming I understood what you meant) I think we should deal with this in `triton::getPointeeBitWidth`?